include/nbl/builtin/hlsl/bxdf/ndf.hlsl

```diff
@@ -0,0 +1,339 @@
+    scalar_type operator()(SAnisotropicParams<scalar_type> params)
+    {
+        scalar_type n = (params.TdotH2 * params.ny + params.BdotH2 * params.nx) * params.one_minus_NdotH2_rcp;
+        return (isinf<scalar_type>(params.nx) || isinf<scalar_type>(params.ny)) ?  numeric_limits<scalar_type>::infinity : 
+            sqrt<scalar_type>((params.nx + 2.0) * (params.ny + 2.0)) * numbers::inv_pi<scalar_type> * 0.5 * pow<scalar_type>(params.NdotH, n);
+    }
+};
+
+template<typename T NBL_PRIMARY_REQUIRES(is_scalar_v<T>)
+struct Beckmann
+{
+    using scalar_type = T;
+
+    scalar_type operator()(SIsotropicParams<scalar_type> params)
+    {
+        scalar_type nom = exp<scalar_type>( (params.NdotH2 - 1.0) / (params.n_or_a2 * params.NdotH2) ); // exp(x) == exp2(x/log(2)) ?
+        scalar_type denom = params.n_or_a2 * params.NdotH2 * params.NdotH2;
+        return numbers::inv_pi<scalar_type> * nom / denom;
+    }
+
+    scalar_type operator()(SAnisotropicParams<scalar_type> params)
+    {
+        scalar_type nom = exp<scalar_type>(-(params.TdotH2 / params.ax2 + params.BdotH2 / params.ay2) / params.NdotH2);
+        scalar_type denom = params.ax * params.ay * params.NdotH2 * params.NdotH2;
+        return numbers::inv_pi<scalar_type> * nom / denom;
+    }
+};
+
+
+template<typename T NBL_PRIMARY_REQUIRES(is_scalar_v<T>)
+struct GGX
+{
+    using scalar_type = T;
+
+    // trowbridge-reitz
+    scalar_type operator()(SIsotropicParams<scalar_type> params)
+    {
+        scalar_type denom = params.NdotH2 * (params.n_or_a2 - 1.0) + 1.0;
+        return params.n_or_a2 * numbers::inv_pi<scalar_type> / (denom * denom);
+    }
+
+    scalar_type operator()(SAnisotropicParams<scalar_type> params)
+    {
+        scalar_type a2 = params.ax * params.ay;
+        scalar_type denom = params.TdotH2 / params.ax2 + params.BdotH2 / params.ay2 + params.NdotH2;
+        return numbers::inv_pi<scalar_type> / (a2 * denom * denom);
+    }
+
+    // burley
+    scalar_type operator()(SAnisotropicParams<scalar_type> params, scalar_type anisotropy)
+    {
+        scalar_type antiAniso = 1.0 - anisotropy;
+        scalar_type atab = params.ax * antiAniso;
+        scalar_type anisoTdotH = antiAniso * params.TdotH;
+        scalar_type anisoNdotH = antiAniso * params.NdotH;
+        scalar_type w2 = antiAniso/(params.BdotH * params.BdotH + anisoTdotH * anisoTdotH + anisoNdotH * anisoNdotH * params.ax);
+        return w2 * w2 * atab * numbers::inv_pi<scalar_type>;
+    }
+};
+
+// common
+namespace impl
+{
+template<class T, class U>
+struct is_ggx : bool_constant<
+    is_same<T, GGX<U> >::value
+> {};
+}
+
+template<class T> 
+struct is_ggx : impl::is_ggx<T, typename T::scalar_type> {};
+
+template<typename T>
+NBL_CONSTEXPR bool is_ggx_v = is_ggx<T>::value;
+
+
+enum MicrofacetTransformTypes : uint16_t
+{
+    REFLECT_BIT = 0b01,
+    REFRACT_BIT = 0b10,
+    REFLECT_REFRACT_BIT = 0b11
+};
+
+template<typename NDF, uint16_t reflect_refract>
```

Comment:
is there still a DXC bug where you can't template on an enum ?

I'll update DXC on master today and you can merge.

Reply:
sorry it took so long, about to be updated

Reply:
i s  there any remaining place where we template on the `uint16_T` instead of `MicrofacetTransformTypes` ?